### ISSUE TITLE
memory leak in kernel space because the memory registered by IPV6_ADD_MEMBERLAND is not released

### ISSUE THREAD

**Issue by gaoxingwang** [radvd-project/radvd#177]
Hello, I found a problem. 
When I use radvd of version 2.18 , after the process restarts or receives the sighup signal for multiple times , all ifaces are released, but only the user space memory is released. The mc memory registered with the ICMPv6 sock through IPV6_ADD_MEMBERSHIP is not released. Maybe I should add IPV6_DROP_MEMBERSHIP here to clean up resources, am I right?

radvd version：2.18
kernel: linux 4.19.90

**Comment by stappersg** [comment]
On Sun, Jun 05, 2022 at 11:34:35PM -0700, gaoxingwang wrote:
> Maybe I should add IPV6_DROP_MEMBERSHIP here to clean up resources, am I right?

Patches (merge requests) will be reviewed.

 
Groeten
Geert Stappers
-- 
Silence is hard to parse

**Comment by robbat2** [comment]
@gaoxingwang this sounds like a kernel bug.

Can you share how you're measuring the kernel mc memory leaking?

The kernel should implicitly drop membership when the socket is closed due to process termination (the kernel should also be idempotent when adding membership, to not continue to grow).

**Comment by gaoxingwang** [comment]
One day, we noticed this problem that the radvd was abnormal and the following alarm was reported:
![image](https://user-images.githubusercontent.com/26615073/172978244-59cfd02b-9284-4629-b340-03dd769b1833.png)

The ftrace command output shows that when the sendmsg returns -ENOBUFS, the function exits from sock_kmalloc, indicating that the memory is insufficient.
![image](https://user-images.githubusercontent.com/26615073/172979467-e5f10401-9151-4132-a934-4915f5055ea6.png)

To find out the reason why the memory is occupied, print the function stack of kprobe and find out that the radvd process invokes setsockopt. The memory increases by 72 bytes each time and the size is 0x48, and the memory will not be released.
![image](https://user-images.githubusercontent.com/26615073/172979238-b460296a-572f-4a1f-8536-0a6c819d5c09.png)

**Comment by robbat2** [comment]
Thanks, that's a fantastic bug report!
